lib/ui/painting.dart

```diff
@@ -1607,12 +1655,152 @@ class Path extends NativeFieldWrapperClass2 {
+  static Path combine(PathOperation operation, Path path1, Path path2) {
+    assert(path1 != null);
+    assert(path2 != null);
+
+    final Path path = new Path();
+    if (path._op(path1, path2, operation.index)) {
+      return path;
+    } 
+    throw new StateError('Path combine failed - likely due to invalid path');
+  }
+  bool _op(Path path1, Path path2, int operation) native 'Path_op';
+
+  /// Creates a [PathMetrics] object for this path.
+  PathMetrics computeMetrics({bool forceClosed = false}) {
+    return new PathMetrics._(this, forceClosed);
+  }
+}
+
+/// Convenience class to return the result of [PathMetric.getTangentForOffset].
+class Tangent {
+  /// Position of the tangent
+  final Offset position;
+  /// Angle of the tangent in radians
+  final double angle;
+
+  const Tangent(this.position, this.angle);
+}
+
+/// An iterable collection of [PathMetric] objects.
+/// 
+/// Each path metric corresponds to a segment of a path.
+/// 
+/// For example, if you call [Path.moveTo] once in the middle of drawing, you
+/// would get two PathMetric objects in this iterable.  Call 
+/// [Path.computeMetrics] to create the iterable.
+class PathMetrics extends collection.IterableBase<PathMetric> {
+  final Iterator<PathMetric> _iterator;
+
+  PathMetrics._(Path path, bool forceClosed) :
+    _iterator = new PathMetricIterator._(new PathMetric._(path, forceClosed));
+
+  @override
+  Iterator<PathMetric> get iterator => _iterator;
+}
+
+/// Tracks iteration from one segment of a path to the next for measurement.
+class PathMetricIterator implements Iterator<PathMetric> {
+  PathMetric _pathMetric;
+  bool _firstTime = true;
+
+  PathMetricIterator._(this._pathMetric);
+
+  @override
+  PathMetric get current => _firstTime ? null : _pathMetric;
+
+  @override
+  bool moveNext() {
+    // PathMetric isn't a normal iterable - it's already initialized to its
+    // first Path.  Should only call _moveNext when done with the first one. 
+    if (_firstTime == true) {
+      _firstTime = false;
+      return true;
+    } else if (_pathMetric?._moveNext() == true) {
+      return true;
+    } else {
+      _pathMetric = null;
+      return false;
+    }
+  }
+}
+
+/// Utilities for measuring a [Path] and extracting subpaths.
+///
+/// Call [Path.computeMetrics] to create these objects. Once created, measures
+/// will only be valid while the path remains unmodified.  If the path is 
+/// modified, the behavior of the PathMetric object is undefined.
+class PathMetric extends NativeFieldWrapperClass2 {
+  /// Create a new empty [Path] object.
+  PathMetric._(Path path, bool forceClosed) { _constructor(path, forceClosed); }
+  void _constructor(Path path, bool forceClosed) native 'PathMeasure_constructor';
+
+  /// Return the total length of the current contour.
+  double get length native 'PathMeasure_getLength';
+
+  /// Pins distance to 0 <= distance <= getLength(), and then computes the 
+  /// corresponding position and tangent.
```

Comment:
```
/// Computes the position of the current contour at the given offset, and the angle of the path at that point.
///
/// For example, calling this method with a distance of 1.4 for a line from 0.0,0.0 to 2.0,2.0 would give
/// the point 1.0,1.0 and the angle 45 degrees (but in radians).
///
/// Returns null if the contour has zero [length].
///
/// The distance is clamped to the length of the current contour.
```

Reply:
also, add a test that checks that if you use this method with the path described in that example, you do get those results.

Reply:
I'll add some tests.  The numbers won't come out quite that nice and even - what do you think of

`For example, calling this method with a distance of 1.41 ... would give a point near 1.0,1.0 and the angle near 0.785 radians (near 45 degrees)`?

Reply:
I don't think it matters if the example is 100% precise. Pedantic precision can get in the way of the message sometimes.